pkg/model/core/stack.go

```diff
@@ -60,8 +79,36 @@ func (s *defaultStack) AddResource(res Resource) error {
 	return nil
 }
 
+func (s *defaultStack) RemoveResource(id graph.ResourceUID) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.resources, id)
+	s.resourceGraph.RemoveNode(id)
+}
+
+func (s *defaultStack) AddService(service *corev1.Service) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.services[string(service.UID)] = *service
+}
+
+func (s *defaultStack) RemoveService(service *corev1.Service) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.services, string(service.UID))
+}
+
+func (s *defaultStack) ListServices() (result []corev1.Service) {
+	for _, service := range s.services {
+		result = append(result, service)
+	}
+	return
```

Comment:
If we're worried that this will be accessed from multiple goroutines, then we need a (read) lock around the reads, too.

Reply:
Thanks. It makes me rethink about the whole process. So for now, stack will hold a lock and will only allow one service to modify/deploy stack at one time. This makes thread safe and reduce the race condition.